File: leetcode/src/main/java/com/giwankim/leetcode/MergeIntervals.java
package com.giwankim.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class MergeIntervals {
  /**
   * @implNote Time {@code O(n log n)}, space {@code O(n)}.
   */
  public int[][] merge(int[][] intervals) {
    List<int[]> result = new ArrayList<>();
    Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
    for (int i = 0; i < intervals.length; i++) {
      if (result.isEmpty() || result.getLast()[1] < intervals[i][0]) {
        result.add(intervals[i]);
        continue;
      }
      result.getLast()[1] = Math.max(result.getLast()[1], intervals[i][1]);
    }
    return result.toArray(new int[0][]);
  }
}

File: leetcode-kt/src/main/kotlin/com/giwankim/leetcode/MergeIntervals.kt
package com.giwankim.leetcode

import kotlin.math.max

class MergeIntervals {
    fun merge(intervals: Array<IntArray>): Array<IntArray> {
        val result = mutableListOf<IntArray>()

        intervals.sortWith(compareBy({ it[0] }, { it[1] }))
        var start = intervals[0][0]
        var end = intervals[0][1]

        for (i in 1 until intervals.size) {
            if (end < intervals[i][0]) {
                result.add(intArrayOf(start, end))
                start = intervals[i][0]
                end = intervals[i][1]
            } else {
                end = max(end, intervals[i][1])
            }
        }
        result.add(intArrayOf(start, end))
        return result.toTypedArray()
    }
}
